ui/src/app/applications/components/applications-sync-panel/applications-sync-panel.tsx

```diff
@@ -1,159 +1,105 @@
 import {services} from '../../../shared/services';
 import {ApplicationRetryOptions} from '../application-retry-options/application-retry-options';
 import {ApplicationManualSyncFlags, ApplicationSyncOptions, FORCE_WARNING, SyncFlags} from '../application-sync-options/application-sync-options';
 import {ApplicationSelector} from '../../../shared/components';
 import {confirmSyncingAppOfApps, getAppDefaultSource} from '../utils';
+import {ApplicationsOperationPanel} from '../applications-operation-panel/applications-operation-panel';
 
 interface Progress {
     percentage: number;
     title: string;
 }
+interface OperationHandlerContext {
+    setProgress: (progress: Progress) => void;
+    ctx: ContextApis;
+}
 
 export const ApplicationsSyncPanel = ({show, apps, hide}: {show: boolean; apps: models.Application[]; hide: () => void}) => {
-    const [form, setForm] = React.useState<FormApi>(null);
-    const [progress, setProgress] = React.useState<Progress>(null);
-    const getSelectedApps = (params: any) => apps.filter((_, i) => params['app/' + i]);
-    const [isPending, setPending] = React.useState(false);
-    const syncHandler = (currentForm: FormApi, ctx: ContextApis, applications: models.Application[]) => {
-        const formValues = currentForm.getFormState().values;
+    const validate = async (formApi: FormApi, ctx: ContextApis) => {
+        const formValues = formApi.getFormState().values;
         const replaceChecked = formValues.syncOptions?.includes('Replace=true');
-        const selectedApps = [];
-        const selectedAppOfApps: models.Application[] = [];
-        let containAppOfApps = false;
+        const selectedApps = apps.filter((_, i) => formValues['app/' + i]);
+        const selectedAppOfApps = selectedApps.filter(app => app.isAppOfAppsPattern);
 
-        for (const key in formValues) {
-            if (key.startsWith('app/') && formValues[key]) {
-                selectedApps.push(applications[parseInt(key.slice(key.lastIndexOf('/') + 1), 10)]);
-            }
+        if (replaceChecked && selectedAppOfApps.length > 0) {
+            const confirmed = await confirmSyncingAppOfApps(selectedAppOfApps, ctx, formApi);
+            return confirmed;
         }
+        return true;
+    };
 
-        selectedApps.forEach(app => {
-            if (app.isAppOfAppsPattern) {
-                containAppOfApps = true;
-                selectedAppOfApps.push(app);
+    const handleSubmit = async (selectedApps: models.Application[], params: any, {ctx, setProgress}: OperationHandlerContext) => {
+        if (params.syncFlags?.Force) {
+            const confirmed = await ctx.popup.confirm('Synchronize with force?', () => (
+                <div>
+                    <i className='fa fa-exclamation-triangle' style={{color: ARGO_WARNING_COLOR}} /> {FORCE_WARNING} Are you sure you want to continue?
+                </div>
+            ));
+            if (!confirmed) {
+                return;
```

Comment:
@linghaoSu , I'm quite impressed with how you have taken the refactoring further than what I had in mind. To me, there is a 'healthy' level of code duplication because 1) it is easier to maintain 2) it is easier to read and understand. Have to strike a balance.

The ordering of the steps have changed.  So, at this point, if you try to enable the Force, submit, and then select Cancel, we're going to be in a bind. Try this out.